### PR TITLE
Feat/DeepAR

### PR DESCRIPTION
**Pull request by pennfranc**
Adds the option to make `RNNModel` probabilistic by setting the `__init__` argument `probabilistic` to `True`, which makes it almost equivalent to the DeepAR model described here: https://arxiv.org/pdf/1704.04110.pdf.

To achieve this I had to:
- move the logic of producing model outputs for training and prediction into two separate functions that can be overridden
- move the logic of computing the loss into a separate function that can be overridden
- let `RNNModel` override all of the above functions
- add a new factory function (`TimeSeries.from_times_and_values_probabilistic`) to create probabilistic time series more easily from numpy arrays
- introduce `num_samples` parameter to `TorchForecastingModel.predict` and adapt method to support probabilistic forecasts
- introduce a `LikelihoodModel` class to abstract different learnt probability distributions

Checkout DeepAR in action here: https://github.com/unit8co/darts/blob/feat/DeepAR/examples/09-DeepAR-examples.ipynb

### CODE REVIEW

File: darts/timeseries.py
Comment (hrzn) [inline darts/timeseries.py:349]: Actually could we re-use `from_times_and_values()` and have it behave differently depending whether the dimensionality of `values` is 3 (like here), 2 (like the current `from_times_and_values() - which we have no reason to keep dependent on DataFrame), or 1 (in which case we assume it's a deterministic series with 1 component).

**Comment by hrzn** [comment]
The approach looks very good! Great job @pennfranc 👍

File: darts/timeseries.py
Comment (pennfranc) [reply darts/timeseries.py]: Good point, done! Doing this I was also able to simplify the `TimeSeries.from_values()` method. Let me know if the changes make sense

File: darts/models/torch_forecasting_model.py
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py:213]: Sounds good but I think we still need an attribute, because the RNN could be probabilistic or not, so I guess we can return the attribute in the abstract method.. would that make sense?

File: darts/models/torch_forecasting_model.py
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py:407]: Our 2 efforts have now combined! Haha

File: darts/models/torch_forecasting_model.py
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py]: Update: I have now added the first version of the likelihood model abstraction (`darts.utils.likelihood_model.LikelihoodModel`), and so I am now using the `self.likelihood_model` attribute instead of `self.is_probabilistic`. Of course we could still create a `is_probabilistic()` method which returns False if `self.likelihood_model==None` and True otherwise. What do you think?

File: darts/timeseries.py
Comment (hrzn) [inline darts/timeseries.py:342]: Sound logic, however, are you maybe mutating the time index provided in argument here?
You could maybe simply store a `time_name` variable here.

File: darts/timeseries.py
Comment (hrzn) [inline darts/timeseries.py:356]: So here you'd be using `time_name` instead of `times.name`

File: darts/timeseries.py
Comment (hrzn) [inline darts/timeseries.py:370]: This is so much better like that! 👍

File: darts/utils/likelihood_models.py
Comment (hrzn) [inline darts/utils/likelihood_models.py:24]: I think you inverted the parameters of `_compute_loss()` and `_sample()`

File: darts/utils/likelihood_models.py
Comment (hrzn) [inline darts/utils/likelihood_models.py:14]: This is perfect, excellent way to represent probabilistic likelihoods. Great job!

File: darts/models/rnn_model.py
Comment (hrzn) [inline darts/models/rnn_model.py:83]: Should we make `num_layers` a parameter?

File: darts/models/rnn_model.py
Comment (hrzn) [inline darts/models/rnn_model.py:154]: Maybe you could say here it expects a `darts.utils.LikelihoodModel` so that users reading the docstring know where to look to instantiate their likelihood.

**Review by hrzn — APPROVED** [review]
Super good stuff!

File: darts/models/rnn_model.py
Comment (pennfranc) [reply darts/models/rnn_model.py:83]: Yes, actually I already did that and it's merged into develop. I'm just waiting for your timeseries refactoring to be merged into develop before merging this branch with the newest version of develop. I'm hoping to avoid some annoying merge conflicts that way

File: darts/utils/likelihood_models.py
Comment (hrzn) [inline darts/utils/likelihood_models.py:35]: It's slightly weird to name it `output`, because even though most of the time it's the output of a network, from the point of view of this module what's received here is more of an input. But this is really a detail, no need to change (unless you really want to).

Comment (pennfranc) [reply darts/utils/likelihood_models.py:35]: True. How about `model_output` or `likelihood_parameters`?

File: darts/models/torch_forecasting_model.py
Comment (hrzn) [inline darts/models/torch_forecasting_model.py:595]: Actually, isn't there a way to "tile" the `input_series` and `cov_future`, in a way that would allow only a simple call to `predict_*_model()` using batching? That would make things faster.

File: darts/models/rnn_model.py
Comment (hrzn) [inline darts/models/rnn_model.py:34]: I think this could be named simply `likelihood`

File: darts/utils/likelihood_models.py
Comment (hrzn) [inline darts/utils/likelihood_models.py:52]: Why `sum` and not `mean` here? By default for the other deterministic models we are using `MSELoss()` which defaults to a `mean` reduction.

Comment (pennfranc) [reply darts/utils/likelihood_models.py:52]: No good reason, I changed it

File: darts/utils/likelihood_models.py
Comment (hrzn) [reply darts/utils/likelihood_models.py:35]: model_output is good 👍

**Pull request merged**
